### PR TITLE
Add new `ballista-cli` crate

### PR DESCRIPTION
**Pull request by andygrove**
# Which issue does this PR close?

Closes https://github.com/apache/arrow-datafusion/issues/2493 and https://github.com/apache/arrow-datafusion/issues/2433 and https://github.com/apache/arrow-datafusion/issues/2517

 # Rationale for this change

- Allows us to release DataFusion and Ballista separately. This was not practical before because the datafusion-cli depended on ballista which in turn depended on datafusion, leading to issues such as https://github.com/apache/arrow-datafusion/issues/2433)
- Allows DataFusion and Ballista CLIs to evolve separately without Ballista adding development overhead to the DataFusion CLI

Here is the new crate dependency diagram:

crate-deps

# What changes are included in this PR?

- [x] Create new ballista-cli crate by copying all of datafusion-cli crate
- [x] Remove ballista features from datafusion-cli crate
- [x] Update dependency diagram
- [x] Refactor to have ballista-cli re-use some of the code from datafusion-cli instead of duplicating all of the code
- [x] Update CI scripts
- [x] Update user guide

# Are there any user-facing changes?

Yes

### CODE REVIEW

File: ballista-cli/src/context.rs [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+//! Context (remote or local)
+
+use datafusion::dataframe::DataFrame;
+use datafusion::error::{DataFusionError, Result};
+use datafusion::execution::context::{SessionConfig, SessionContext};
+use std::sync::Arc;
+
+/// The CLI supports using a local DataFusion context or a distributed BallistaContext
+pub enum Context {
```
Comment (alamb) [inline ballista-cli/src/context.rs:26]: this can probably be simplified over time as this cli is always for ballista

**Review by alamb — APPROVED** [review]

**Comment by alamb** [comment]
Thanks @andygrove  looks much better to me

File: ballista-cli/src/context.rs (hunk 1 above)
Comment (andygrove) [reply ballista-cli/src/context.rs:26]: Agreed. I think ballista still needs a local option but it should be using a local (in-process) ballista scheduler and executor and not using datafusion directly.

**Comment by andygrove** [comment]
Thanks for the review @alamb. I pushed another commit to fix the `datafusion-cli` lock file that had conflicts after merging from master and also added the `ballista-cli` lock file that was previously ignored. I also updated the CI integration test to build the `ballista-cli`.

**Comment by andygrove** [comment]
AMD64 tests failed with `No space left on device` ... trying again

**Comment by andygrove** [comment]
AMD64 tests failed with `No space left on device` 3 times in a row now.

**Comment by andygrove** [comment]
@alamb All PR builds are failing on AMD64 with `no space left on device`. Any idea what we need to do to resolve?

**Comment by andygrove** [comment]
> > 
> 
> @alamb All PR builds are failing on AMD64 with no space left on device. Any idea what we need to do to resolve?
> 
> Perhaps we can try to update the cache-key in https://github.com/apache/arrow-datafusion/blob/master/.github/workflows/rust.yml#L55 to see if something about the old caches is too large?

Thanks. I suppose this cache just continues to grow forever and doesn't clear out old versions of deps that we no longer use? Perhaps we should include the month name in the cache so it resets periodically?

**Comment by andygrove** [comment]
@alamb Moving the ballista tests to a separate job fixes this. Ok to merge?

File: .github/workflows/rust.yml [hunk 2]
```diff
@@ -123,9 +115,9 @@ jobs:
         run: |
           export ARROW_TEST_DATA=$(pwd)/testing/data
           export PARQUET_TEST_DATA=$(pwd)/parquet-testing/data
+          cargo test
```
Comment (andygrove) [inline .github/workflows/rust.yml:118]: We were previously only running `cargo test` in the `datafusion-examples` crate in this step

File: .github/workflows/rust.yml [hunk 3]
```diff
@@ -75,14 +75,6 @@ jobs:
         env:
           CARGO_HOME: "/github/home/.cargo"
           CARGO_TARGET_DIR: "/github/home/target"
-      - name: Check DataFusion CLI builds with ballista
-        run: |
-          (cd datafusion-cli && cargo check)
-          (cd datafusion-cli && cargo check --no-default-features)
-          (cd datafusion-cli && cargo check --features=ballista)
```
Comment (andygrove) [inline .github/workflows/rust.yml:82]: We test `datafusion-cli` in the integration tests and we no longer support running this with ballista

**Pull request merged**